### PR TITLE
[PPWM-77] Add Persistent Disk Notification to Workspace List Page

### PR DESCRIPTION
**Pull request by ehigham**
RR: https://broadworkbench.atlassian.net/browse/PPWM-77 
Add a notification to the workspace list page if the user has any persistent disks that need migrating (ie created in the workspaces old google project).

Notes
- this notification appears in the workspace list page rather than the main landing page to avoid loading all workspaces and disks before the user does anything.
- notification text from https://github.com/DataBiosphere/terra-ui/pull/3425 

image

<!--
Hello, friend!
Remember to mention what sort of testing/verification you performed in the course of building this PR, i.e. checking functionality in the browser locally.
Also, so if a screen recording [1] and/or screenshots would be a helpful way to communicate context, please consider including some in your PR description.
Thanks!

[1] https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac
--!>

### CODE REVIEW

File: src/pages/workspaces/List.js [hunk 1]
```diff
@@ -91,6 +94,18 @@ const workspaceSubmissionStatus = workspace => {
   )
 }
 
+const anyPersistentDiskRequiresMigration = (workspaces, disks) => {
+  const workspacesByNamespace = _.flow(
```
Comment (jdcanas) [inline src/pages/workspaces/List.js]: maybe a comment here on exactly how this transforms the data? its a bit hard to follow with all the functions invoked...

**Review by jdcanas — APPROVED** [review]
lgtm, maybe just a comment on example input and what the data looks at important parts of the transformation where I noted

File: src/pages/workspaces/List.js [hunk 2]
```diff
@@ -128,12 +143,24 @@ export const WorkspaceList = () => {
...
+  const [gcpPersistentDisks, setGcpPersistentDisks] = useState()
+
   useOnMount(() => {
     const loadFeatured = async () => {
       setFeaturedList(await Ajax().FirecloudBucket.getFeaturedWorkspaces())
     }
 
+    const loadPersistentDisks = async () => {
+      setGcpPersistentDisks(_.filter(({ cloudContext }) => isGcpContext(cloudContext),
+        await Ajax().Disks.list({
```
Comment (cahrens) [inline src/pages/workspaces/List.js:158]: You need error handling around this in case the Ajax call fails.

There should be a ticket to rip this code out once the migration period is over and a TODO in the code. We don't want to keep the loading of the disks in this file permanently.

Comment (ehigham) [reply src/pages/workspaces/List.js:158]: shouldn't the other request have error handling too?

Comment (cahrens) [reply src/pages/workspaces/List.js:158]: Yes

Comment (cahrens) [reply src/pages/workspaces/List.js:158]: With `getFeauredWorkspaces` we probably want to use `withErrorIgnoring` though since we wouldn't want to tell the user if loading the workspaces failed (and it may be expected in some environments).

File: src/pages/workspaces/List.js [hunk 3]
```diff
@@ -499,7 +526,13 @@ export const WorkspaceList = () => {
         workspace: getWorkspace(requestingAccessWorkspaceId),
         onDismiss: () => setRequestingAccessWorkspaceId(undefined)
       }),
-      loadingWorkspaces && (!workspaces ? transparentSpinnerOverlay : topSpinnerOverlay)
+      loadingWorkspaces && (!workspaces ? transparentSpinnerOverlay : topSpinnerOverlay),
+      !loadingWorkspaces && anyPersistentDiskRequiresMigration(workspaces, gcpPersistentDisks) && h(UnboundDiskNotification, {
```
Comment (cahrens) [inline src/pages/workspaces/List.js]: There doesn't appear to be any way to get keyboard focus into this notification to interact with the link.

Comment (ehigham) [reply src/pages/workspaces/List.js]: OK, how would I do that?

Comment (cahrens) [reply src/pages/workspaces/List.js]: I would recommend that you just put it directly inline in the page. Perhaps under "Dedicated spaces for you and…"?

Comment (ehigham) [reply src/pages/workspaces/List.js]: I'm following the UX mockup for the banner in the ticket. Are you suggesting not doing that?

File: src/pages/workspaces/List.js [hunk 4]
```diff
@@ -499,7 +525,13 @@ export const WorkspaceList = () => {
...
```
Comment (cahrens) [inline src/pages/workspaces/List.js]: It is possible that the workspaces Ajax response has returned, but the persistent disk one has not. I would recommend changing `anyPersistentDiskRequiresMigration` so that it simply returns false if either of `workspaces` or `gcpPersistentDisks` is undefined. Then you can always call it without concern.

File: src/pages/workspaces/List.js (hunk 3 above)
Comment (cahrens) [reply src/pages/workspaces/List.js]: It has to be keyboard accessible, and given the relatively small audience and the fact that this feature will be short-lived, I was offering a suggestion that I thought would be the simplest to implement.

But certainly it is worth checking with UX to see if they are OK with it being inline, if you don't wish to figure out how to get the notification into the page's natural tab ordering.

Comment (cahrens) [reply src/pages/workspaces/List.js]: It might be enough to just move it up in the HTML. Perhaps if I could get to the end of the workspaces list it would actually then go into the notification?

Comment (cahrens) [reply src/pages/workspaces/List.js]: Ah yes, it goes into the notification after navigating through the last workspace in the list. So insert the HTML up at the top of the page.

File: src/pages/workspaces/List.js [hunk 5]
```diff
@@ -95,15 +95,15 @@ const workspaceSubmissionStatus = workspace => {
...
+const somePersistentDiskRequiresMigration = (workspaces, disks) => {
   // construct map of namespace -> (map of name -> workspace) from list of workspace
   const workspacesByNsAndName = _.flow(
     _.map('workspace'),
     _.groupBy('namespace'),
     _.mapValues(_.flow(_.groupBy('name'), _.mapValues(_.head)))
   )(workspaces)
 
-  return _.flip(_.some)(disks, ({ googleProject, labels }) => {
+  return workspaces && _.flip(_.some)(disks, ({ googleProject, labels }) => {
```
Comment (cahrens) [inline src/pages/workspaces/List.js:106]: So this works OK if `disks` is undefined?

Comment (ehigham) [reply src/pages/workspaces/List.js:106]: yes  - you can verify this yourself in the console: 
```javascript
> _.some(() => true, undefined)
false
```

File: src/pages/workspaces/List.js (hunk 3 above)
Comment (ehigham) [reply src/pages/workspaces/List.js]: @tiffanytuffany

Comment (ehigham) [reply src/pages/workspaces/List.js]: Where should the banner go then? after the top bar? in the "main" hierarchy?

**Review by cahrens — APPROVED** [review]
